pkg/orderers/pbftgoodcase.go

```diff
@@ -190,6 +191,36 @@ func (orderer *Orderer) applyMsgPreprepare(preprepare *ordererspbftpb.Preprepare
 		)
 	}
 
+	// Future view changes decide empty segments, no need to have a certificate
+	if orderer.view > 0 && preprepare.Aborted && len(preprepare.CertData) == 0 {
+		return orderer.finishUpPreprepare(preprepare, slot)
+	}
+
+	// append to FIFO
+	orderer.notVerifiedPrepepareContext = append(orderer.notVerifiedPrepepareContext, &verifyCertContext{
+		preprepare,
+		slot,
+	})
+
+	//Verify certificate from preprepare message before saving it in the slot
+	return orderer.verifyCert(preprepare) // TODO Alejandro: change to verify certs in the slice (new message? probs best?)
+
+}
+
+// Finalize applyMsgPreprepare after cert verified
+func (orderer *Orderer) applyCertVerified(verified *availabilitypb.CertVerified) (*events.EventList, error) {
+	// TODO Alejandro: change to verify certs in the slice (argument is now slice)
+	//pop from FIFO
+	context := orderer.notVerifiedPrepepareContext[0]
+	orderer.notVerifiedPrepepareContext = orderer.notVerifiedPrepepareContext[1:]
+
+	if !verified.Valid {
+		return events.EmptyList(), fmt.Errorf(verified.Err)
+	}
+	return orderer.finishUpPreprepare(context.prepepare, context.slot), nil
+}
+
+func (orderer *Orderer) finishUpPreprepare(preprepare *ordererspbftpb.Preprepare, slot *pbftSlot) *events.EventList {
```

Comment:
In fact, I think that in our particular case it is better to only check the validity of the certificates after agreeing on them. It is better for ISS finish the agreement on a certificate, even if it's invalid, and discard it later, then to wait for a timeout.

I suggest removing this check and addressing it in a separate PR.

Another reason is that it will make merging easier, as this code conflicts with the changes I made to implement the agreement on checkpoint certificates.

Reply:
Done, left comment to be addressed by future PR (once merge occurs). Happy to address it then.